examples/opinion_spread.jl

```diff
@@ -0,0 +1,92 @@
+# # Opinion spread
+
+# This is a simple model of how an opinion spreads through a community. Each individual has a number of opinions as a list of integers. They can change their opinion by changing the numbers in the list.
+
+# Agents can change their opinion at each step. They choose one of their neighbors randomly, and adopt one of the neighbor's opinion. They are more likely to adopt their neighbors opinion if the share more opinions with each other.
+
+using Agents
+using AgentsPlots
+using Random
+
+# ## Building the model
+# ### 1. Model creation
+
+mutable struct Citizen <: AbstractAgent
+	id::Int
+	pos::Tuple{Int, Int}
+	stabilized::Bool
+	opinion::Array{Int, 1}
+	prev_opinion::Array{Int, 1}
+end
+
+function create_model(;dims=(10, 10), nopinions=3, levels_per_opinion=4)
+	space = GridSpace(dims, periodic=true, moore=true)
+	properties = Dict(:nopinions=>nopinions) 
+	model = AgentBasedModel(Citizen, space, scheduler=random_activation, properties=properties)
+	for cell in 1:nv(model)
+		add_agent!(cell, model, false, rand(1:levels_per_opinion, nopinions), rand(1:levels_per_opinion, nopinions))
+	end
+	return model
+end
+
+# ### 2. Stepping functions
+
+function adopt!(agent, model)
+	neighbor = rand(space_neighbors(agent, model))
+	matches = model[neighbor].opinion .== agent.opinion
+	nmatches = count(matches)
+	# Adopt a different opinion w/ calculated probability
+	if nmatches < model.nopinions && rand() < nmatches/model.nopinions
+		switchId = rand(findall(x-> x==false, matches))
+		agent.opinion[switchId] = model[neighbor].opinion[switchId]
+	end
+end
+
+function update_prev_opinion!(agent, model)
+	for i in 1:model.nopinions
+		agent.prev_opinion[i] = agent.opinion[i]
+	end
+end
+
+function is_stabilized!(agent, model)
+	if agent.prev_opinion == agent.opinion
+		agent.stabilized = true
+	else
+		agent.stabilized = false
+	end
+end
+
+function agent_step!(agent, model)
+	update_prev_opinion!(agent, model)
+	adopt!(agent, model)
+	is_stabilized!(agent, model)
+end
+
+# ## Running the model
+
+levels_per_opinion=4
+model = create_model(nopinions=3,levels_per_opinion=levels_per_opinion)
+
+"run the model until all agents stabilize"
+rununtil(model, s) = count(getproperty.(values(model.agents), :stabilized)) == prod(model.space.dimensions)
+
+agentdata, _ = run!(model, agent_step!, dummystep, rununtil, adata=[(:stabilized, count)])
```

Comment:
I think this highlights a gap in our `run!` constructors. We state that `model_step!` is optional, so `run!(model, agent_step!)` is possible, but when we want to overload `step`, it's necessary to use `dummystep` explicitly. Would be nice if we can manage to remove this.

Reply:
#284

Reply:
`model_step!` is only optional in the case of `n` being an integer. If `n` is a function `model_step` is mandatory. I've changed this a while back in the docs as well. Did I forget to update the docs of `run!`? Because the docs of `step!` are clear about this.

Reply:
@Libbum see my comment above. I don't think #284 is necessary.

Reply:
Oh, then why is #284 possible? But yeah, `run!(model, agent_step! [, model_step!], n; kwargs...)` is what's listed currently.

Reply:
Technically it isn't. Your method `run!(model, agenstep, modelstep)` conflicts with `run!(model, agentstep, n)` in the case `n` is a function.